prov/sm2/src/sm2_coordination.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023 Amazon.com, Inc. or its affiliates.
+ * All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _SM2_COORDINATION_H_
+#define _SM2_COORDINATION_H_
+
+#include "config.h"
+
+#include <stddef.h>
+#include <stdint.h>
+#include <sys/un.h>
+
+#include <ofi_atom.h>
+#include <ofi_hmem.h>
+#include <ofi_mem.h>
+#include <ofi_proto.h>
+#include <ofi_rbuf.h>
+#include <ofi_tree.h>
+
+#include <rdma/providers/fi_prov.h>
+
+#define SM2_DIR "/dev/shm/"
+#define SM2_NAME_MAX 256
+#define SM2_PATH_MAX (SM2_NAME_MAX + sizeof(SM2_DIR))
+#define SM2_COORDINATION_FILE "/dev/shm/fi_sm2_mmaps"
+#define SM2_COORDINATION_DIR "/dev/shm"
+#define SM2_COORDINATOR_MAX_TRIES 1000
+#define SM2_COORDINATOR_MAX_UNIVERSE_SIZE 4096
+
+struct sm2_mmap {
+	char *base;
+	size_t size;
+	int fd;
+};
+
+struct sm2_private_aux {
```

Comment:
I am working to address several of these comments.  I've read through them, but I think it's worth I clarify something here to make sure I'm not missing anything.

SM2 is an FI_AV_TABLE, with one exception.  fi_addr_t's are assigned sequentially but their scope is not a single AV, but rather all fabrics using sm2 as the same user on the same system.  So instead of an AV_TABLE, we claim AV_MAP.

So back to your your comment "The fiaddrs are going to change depending on the EP".  This is not true in sm2.  The "Coordinator" is in fact coordinating on AV insert to generate these sequential, globally unique IDs.  This is to help out the Nemesis algorithm later, because Nemesis has to be able to address memory of peers that may never added to our AV.  

In our testing we've found that fi_av_map is indeed supported by open mpi, but might have a performance penalty.  I didn't expect the performance penalty so I didn't attempt to make a translation layer to fi_av_table.  I do think we have some pretty straight-forward approaches to address this, probably using the "auxiliary" structure, but you are suggesting we remove the sm2_aux struct in favor of util_av.

I am trying to understand how we could best implement this with util av and it's not making sense to me.   I'll admit some of that is because I'm not familiar with the util_av and I'm having trouble finding documentation on it (since it's internal).

Would you mind describing how you would use the util_av to store and retrieve the global id?

Thanks

Reply:
Yeah totally. First of all, you're totally right in the sense that the above comment (my original one about the EP fiaddrs being different) is incorrect when you use FI_AV_MAP :) I think I reviewed that before I saw you changed the av type.
We can choose to change it to AV_MAP and what you're saying above would be correct though I still don't think you would need the aux struct since the fiaddr would just be equivalent to the global index or some pointer to the info. There's no reason to maintain a backwards lookup since the fiaddr is the internal pointer or index. That's the whole point of FI_AV_MAP is that the provider doesn't have to maintain an index lookup but the application has to instead. This is really useful for more complicated addressing but since the global_id is already indexed, I don't think it's necessary in this use case and would just add overhead and potential limitations to the application. FI_AV_TABLE allows the application to directly use fi_addrs as index lookups. When you make the application use a map, it requires them to do some translation between the fi_addr and their addressing. Therefore FI_AV_TABLE is recommended and I think the implementation you have here is really close to supporting it. I think you'll need the aux struct in the av (though I would remove the struct in general and just make it a simple array allocated in av_init) as you're using it. This will let you go from global_id back to the fiaddr in order to report rx completions. On the send side, you can use the util_av to go from fiaddr to the global_id. Basically you're just using 2 arrays fiaddr[global_id] and global_id[fiaddr]. The first of those arrays is the one currently in the util_av and the second one is what you have with the aux struct. Alternatively, you can get rid of the util_av and just use those 2 arrays if that makes more sense to you.

Basically, what I'm saying is, this implementation is really close to FI_AV_TABLE so unless there's a huge reason to move to FI_AV_MAP, I would suggest leaving it at FI_AV_TABLE so we don't impose that limitation and overhead to the app.

Reply:
Switched to use util_av + an array of FI_ADDRs (removed struct sm2_private_aux)